CustomAudioFileCapturer/CustomCapturerAudioDevice.cs

```diff
@@ -0,0 +1,315 @@
+            }
+        }
+
+        public int GetEstimatedAudioCaptureDelay()
+        {
+            lock (audioDeviceLock)
+            {
+                return 0;
+            }
+        }
+
+        public AudioDeviceSettings GetAudioCapturerSettings()
+        {
+            lock (audioDeviceLock)
+            {
+                return new AudioDeviceSettings();
+            }
+        }
+
+        public void InitAudioRenderer()
+        {
+            lock (audioDeviceLock)
+            {
+                if (IsAudioRendererInitialized())
+                    return;
+                rendererSettings = new AudioDeviceSettings()
+                {
+                    NumChannels = NUMBER_OF_CHANNELS,
+                    SamplingRate = SAMPLING_RATE
+                };
+                isAudioRendererInitialized = true;
+            }
+        }
+
+        public void DestroyAudioRenderer()
+        {
+            lock (audioDeviceLock)
+            {
+                StopAudioRenderer();
+                if (!IsAudioRendererInitialized())
+                    return;
+                isAudioRendererInitialized = false;
+            }
+        }
+
+        public void StartAudioRenderer()
+        {
+            lock (audioDeviceLock)
+            {
+                if (IsAudioRendererStarted())
+                    return;
+                if (!IsAudioRendererInitialized())
+                    throw new Exception("Audio renderer not initialized");
+
+                isAudioRendererStarted = true;
+
+                _ = Task.Factory.StartNew(() =>
+                {
+                    int numberOfSamplesPer10ms = rendererSettings.SamplingRate / 100;
+                    byte[] buffer = new byte[numberOfSamplesPer10ms * rendererSettings.NumChannels * BYTES_PER_SAMPLE];
+                    unsafe
+                    {
+                        fixed (byte* bufferPtr = buffer)
+                        {
+                            DateTime nextBatchTime = DateTime.Now + new TimeSpan(0, 0, 0, 0, 10);
+                            while (isAudioRendererStarted)
+                            {
+                                try
+                                {
+                                    if (DateTime.Now >= nextBatchTime)
+                                    {
+                                        nextBatchTime += new TimeSpan(0, 0, 0, 0, 10);
+                                        int samples = audioBus.ReadRenderData((IntPtr)bufferPtr, numberOfSamplesPer10ms);
+                                        if (isRecording)
+                                        {
+                                            outputFileStream.Write(buffer, 0, numberOfSamplesPer10ms * rendererSettings.NumChannels * BYTES_PER_SAMPLE);
+                                            outputFileStream.Flush();
+                                        }
+                                    }
+                                    Thread.Sleep(1);
```

Comment:
Maybe I already asked in the past. But why don't we sleep until the `nextBatchTime`?

Reply:
In my past experience I've seen that with windows is pretty bad at timing things. If you just sleep(10) this will cause  an accumulated delay by default (and on occasion it will also cause arbitrarily high delays).
Maybe it''d be a good approach to sleep(X) where X is the amount of time remaining to the next batch.
Let me try that. I think it'd look better in a sample.
It's still not what I'd do in a windows app in which timing must be strict but I think here we can do ok with that.

Reply:
Thanks. Yes, now I remember you commented this long time ago.